agent/ae/ae.go

```diff
@@ -47,100 +55,132 @@ type StateSyncer struct {
+// between local and remote servers.
+func (s *StateSyncer) Run() {
+	stagger := func(d time.Duration) time.Duration {
+		f := scaleFactor(s.ClusterSize())
+		return lib.RandomStagger(time.Duration(f) * d)
 	}
-}
 
-// antiEntropy is a long running method used to perform anti-entropy
-// between local and remote state.
-func (ae *StateSyncer) Run() {
-SYNC:
-	// Sync our state with the servers
+FullSync:
 	for {
-		err := ae.State.SyncFull()
+		// attempt a full sync
+		err := s.State.SyncFull()
+
+		// full sync successful, now wait for changes
 		if err == nil {
 			break
 		}
-		ae.Logger.Printf("[ERR] agent: failed to sync remote state: %v", err)
+
+		s.Logger.Printf("[ERR] agent: failed to sync remote state: %v", err)
+
+		// full sync failed.
 		select {
-		case <-ae.ConsulCh:
-			// Stagger the retry on leader election, avoid a thundering heard
+
+		// consul server added to cluster.
+		// retry sooner than retryFailIntv to converge cluster sooner
+		// but stagger delay to avoid thundering herd
+		case <-s.ServerUpCh:
 			select {
-			case <-time.After(lib.RandomStagger(aeScale(syncStaggerIntv, ae.ClusterSize()))):
-			case <-ae.ShutdownCh:
+			case <-time.After(stagger(serverUpIntv)):
+			case <-s.ShutdownCh:
 				return
 			}
-		case <-time.After(syncRetryIntv + lib.RandomStagger(aeScale(syncRetryIntv, ae.ClusterSize()))):
-		case <-ae.ShutdownCh:
+
+		// retry full sync after some time
+		// todo(fs): why don't we use s.Interval here?
+		case <-time.After(retryFailIntv + stagger(retryFailIntv)):
+
+		case <-s.ShutdownCh:
 			return
 		}
 	}
 
-	// Force-trigger AE to pickup any changes
-	ae.changeMade()
-
-	// Schedule the next full sync, with a random stagger
-	aeIntv := aeScale(ae.Interval, ae.ClusterSize())
-	aeIntv = aeIntv + lib.RandomStagger(aeIntv)
-	aeTimer := time.After(aeIntv)
+	// Force-trigger sync to pickup any changes
+	s.triggerSync()
 
 	// Wait for sync events
 	for {
 		select {
-		case <-aeTimer:
-			goto SYNC
-		case <-ae.TriggerCh:
-			// Skip the sync if we are paused
-			if ae.Paused() {
+		// todo(fs): why don't we honor the ServerUpCh here as well?
+		// todo(fs): by default, s.Interval is 60s which is >> 3s (serverUpIntv)
+		// case <-s.ServerUpCh:
+		// 	select {
+		// 	case <-time.After(stagger(serverUpIntv)):
+		// 		continue FullSync
+		// 	case <-s.ShutdownCh:
+		// 		return
+		// 	}
+
+		case <-time.After(s.Interval + stagger(s.Interval)):
+			goto FullSync
+
+		case <-s.TriggerCh:
+			if s.Paused() {
 				continue
 			}
-			if err := ae.State.SyncChanges(); err != nil {
-				ae.Logger.Printf("[ERR] agent: failed to sync changes: %v", err)
+			if err := s.State.SyncChanges(); err != nil {
+				s.Logger.Printf("[ERR] agent: failed to sync changes: %v", err)
 			}
-		case <-ae.ShutdownCh:
+
+		case <-s.ShutdownCh:
 			return
 		}
 	}
 }
+
+// Pause temporarily disables partial sync runs.
+func (s *StateSyncer) Pause() {
+	atomic.AddInt32(&s.paused, 1)
```

Comment:
I think this is defensive programming but I agree that this looks like overkill.